src/main/java/org/swisspush/redisques/util/DefaultRedisProvider.java

```diff
@@ -0,0 +1,85 @@
+package org.swisspush.redisques.util;
+
+import io.vertx.core.*;
+import io.vertx.redis.client.Redis;
+import io.vertx.redis.client.RedisAPI;
+import io.vertx.redis.client.RedisOptions;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Default implementation for a Provider for {@link RedisAPI}
+ *
+ * @author https://github.com/mcweba [Marc-Andre Weber]
+ */
+public class DefaultRedisProvider implements RedisProvider {
+
+    private final Vertx vertx;
+    private RedisquesConfigurationProvider configurationProvider;
+
+    private RedisAPI redisAPI;
+
+    private final AtomicReference<Promise<RedisAPI>> connectPromiseRef = new AtomicReference<>();
+
+    public DefaultRedisProvider(Vertx vertx, RedisquesConfigurationProvider configurationProvider) {
+        this.vertx = vertx;
+        this.configurationProvider = configurationProvider;
+    }
+
+    @Override
+    public Future<RedisAPI> redis() {
+        if(redisAPI != null) {
+            return Future.succeededFuture(redisAPI);
+        } else {
+            return setupRedisClient();
+        }
+    }
+
+    private Future<RedisAPI> setupRedisClient(){
+        Promise<RedisAPI> currentPromise = Promise.promise();
+        Promise<RedisAPI> masterPromise = connectPromiseRef.accumulateAndGet(
+                currentPromise, (oldVal, newVal) -> (oldVal != null) ? oldVal : newVal);
+        if( currentPromise == masterPromise ){
+            // Our promise is THE promise. So WE have to resolve it.
+            connectToRedis().onComplete(event -> {
+                if(event.failed()) {
+                    currentPromise.fail(event.cause());
```

Comment:
Depending on our intent we need to reset the promise in case of failure too. Because as written now, once failed, it is stuck in that failed state forever and won't recover as far I see.

That said, if this is the intended behavior, we're already fine. If not, we may clear the reference so a later call will trigger another retry (eg reset independent of the outcome).

```patch
+  connectPromiseRef.getAndSet(null);
   if( event.failed() ) {
       currentPromise.fail(event.cause());
    } else {
-       connectPromiseRef.getAndSet(null);
        redisAPI = event.result();
        currentPromise.complete(redisAPI);
    }
```